### ISSUE TITLE
[Idea] java executable to directly support (statically compiled) native binaries on all operating systems (as this issue request is an idea, it can be closed easily)

### ISSUE THREAD

**Issue by rubyFeedback** [oracle/graal#7770]
So before I expand on this, let me say that this is just an idea. Please feel free to close it at any moment
in time - I just wanted to give a bit of this idea aka "perhaps this idea may be useful for GraalVM in
the future".

The idea is to basically support creating statically compiled native binary images on all platforms
as-is, via the standard java.exe (rather, java executable).

So, say that I have this ruby .rb file called test.rb:

    def test
      puts 'Hello world! 2+2 = '(2+2).to_s # Don't worry about this being rubbish, it is just a test.
    end

I also have some some .java file that basically calls that .rb file and calls the test() method on it.
I'll skip the content of that .java file, because I just want to illustrate the point.

Ok, so we have a .rb file and a .java file. (We can have a python .py file or anything else, I just
want to demonstrate the idea.)

So, I then invoke java executable via:

    java ./test.rb --static --native-binary

And after some time, java creates a standalone .exe file (or any other executable for that
target operating system). I can then invoke it and it prints e. g. "Hello world! 2 + 2 = 4".

Ok. So far I hope it is quite simple to understand the idea here. The idea is that java
itself, directly, supports the creation of binaries like that.

Internally I don't quite care how this is implemented; perhaps this functionality
requires some add-ons to be downloaded, in which case java can generate a
message that either gets people to download it before running the above; or
automatically downloads it. Either way let's assume here that ALL the required
code is available, and that creating such a native binary works (on Linux that
seemed to work for me when I tried it; I think on windows it may not quite work
right now but who knows - I suppose eventually this may work).

Wouldn't this be a great idea?

Now you may ask "but what is the use case, aside from the already existing
way"?

I am glad you asked!

Basically the above is just an idea of SIMPLIFICATION, that is, rather than us
having to memorize which commands are to be used, we tell the java binary
itself what we want, aka:

"java, groovy graalVM thing, please put these ruby or python files together,
as well as that java bridge code and then generate a statically compiled
super fast executable, without asking me what commandline options I
have to use".

Now I can work around having to memorize these things because I
document stuff locally into files, and I wrote helper scripts that I improve,
e. g. ruby automatically calling the right options and such - but it would
be so much easier and greater if the java binary itself would support
the above. Again, here we assume all things work already as-is, which
may not be the case, but I am speaking here "say, in 2-3 years from
now on" or something like that. Kind of like "universal" binaries.

With "all operating systems" I kind of mean windows, linux and OSX,
although of course the more platforms java can support here, the
better.

Reason why I ask this to be integrated into java executable itself is
for simplicity; but it can also be an "add-on" where people can 
extend java. Which policies can be used is up to others, but here
I assume for use cases of developers, so not necessarily "average
Joe". Simply people who want to create executables in a very
easy way.

Anyway. I think that's my rough idea for now, so let's add a TL;DR and
remember this is just an idea, please feel free to close it - hopefully
some GraalVM dev can read it and comment on it before closing
though, so that we know whether this idea is off the track or may
be good to have (in, again, say 2-3 years from now on; I don't ask
anyone to work on that as-is, I am more interested in the idea 
presentation **and** simplification):

TL;DR:

"Consider having the java executable itself support an easy generation
of statically compiled, native binary, via one, or more, commandline
flags on as many operating platforms as possible, by default."

**Comment by rubyFeedback** [comment]
Hmm this is not a bug ... not sure why that label above ^^^ got auto-added, I may have chosen the wrong template...

**Comment by fernando-valdez** [comment]
The idea of making Java executables support creating native binaries easily across all operating systems is really interesting. It's a big idea with some technical challenges, but it's great to think about for GraalVM's future. Your input is really valuable to us and the community. 
Thanks again for thinking ahead and sharing your thoughts!
